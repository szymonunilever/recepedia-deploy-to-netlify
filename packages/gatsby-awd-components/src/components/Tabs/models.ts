import { ReactComponentElement } from 'react';
import { titleLevel, UnileverLibraryComponent } from '../../models';
import { Tab } from './partials';

export interface TabsProps
  extends UnileverLibraryComponent<AppContent.Tabs.Content> {
  children: ReactComponentElement<typeof Tab>[];
  tabsHeaderContent?: TabsHeaderContent;
  tabFromLocation?: boolean;
  location?: Location;
  data?: TabsData[];
  dataFetched?: boolean;
  titleLevel?: titleLevel;
  results?: number;
}

export interface TabsData {
  pattern: string;
  replacement: string | number;
}

export interface TabsHeaderContent {
  contents: HeaderContent[];
}

export interface HeaderContent {
  heading: string;
  subheading: string;
  view: string;
  defaultSubheading?: string;
}
