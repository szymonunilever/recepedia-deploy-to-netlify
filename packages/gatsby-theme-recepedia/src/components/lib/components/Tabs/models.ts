import { ReactComponentElement } from 'react';
import {
  titleLevel,
  UnileverLibraryComponent,
} from '../../models/globalModels';
import { Tab } from './partials';

export interface TabsProps
  extends UnileverLibraryComponent<AppContent.Tabs.Content> {
  children: ReactComponentElement<typeof Tab>[];
  tabsHeaderContent?: TabsHeaderContent;
  tabFromLocation?: boolean;
  location?: Location;
  data?: TabsData[];
  titleLevel?: titleLevel;
}

export interface TabsData {
  pattern: string;
  replacement: string | number;
}

export interface TabsHeaderContent {
  contents: HeaderContent[];
}

export interface HeaderContent {
  heading: string;
  subheading: string;
  view: string;
  defaultSubheading?: string;
}
