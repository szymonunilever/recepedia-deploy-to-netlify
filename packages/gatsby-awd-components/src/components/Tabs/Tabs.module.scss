@import '../../scss/config';

.tabs {
  position: relative;
  &__buttons {
    display: flex;
    background: $tabs-background;
    height: $tabs-height;
    justify-content: space-around;
  }
  &__button {
    position: relative;
    padding: $tab-indent;
    width: $tab-width;
    background: none;
    border: $tab-border;
    cursor: pointer;
    font-size: $tab-fz;
    font-family: $rubik;
    font-weight: $tab-fw;
    color: $tab-color;
    text-align: $tab-alignment;
    &:before {
      content: '';
      position: absolute;
      bottom: 0;
      left: 0;
      width: 100%;
      height: $tab-line-height;
      background: $tab-line-color;
      transition: width $theme-transition;
    }
    &[aria-selected='false'] {
      &:before {
        width: 0;
      }
    }
    @media (hover: hover),
      (-ms-high-contrast: none),
      (-ms-high-contrast: active) {
      &:hover {
        &:before {
          width: 100%;
        }
      }
    }
  }
}
