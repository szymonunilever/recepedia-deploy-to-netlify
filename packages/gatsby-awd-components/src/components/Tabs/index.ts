export { default, Tabs } from './Tabs';
export { Tab } from './partials';
export * from './models';
