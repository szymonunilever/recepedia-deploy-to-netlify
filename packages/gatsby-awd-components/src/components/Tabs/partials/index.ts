export { Tab } from './Tab';
export * from './models';
