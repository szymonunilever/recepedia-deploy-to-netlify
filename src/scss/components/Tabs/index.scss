@import 'tabs';
@import 'favorites.tabs';
