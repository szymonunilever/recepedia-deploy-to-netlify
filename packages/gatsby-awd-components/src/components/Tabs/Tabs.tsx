import React, { useState, useEffect } from 'react';
import { TabsProps, HeaderContent } from './models';
import theme from './Tabs.module.scss';
import cx from 'classnames';
import Button, { ButtonViewType } from '../Button';
import { Tab } from './partials';
import { TagName, Text } from '../Text';
import get from 'lodash/get';
import getComponentDataAttrs from '../../utils/getComponentDataAttrs';

export const Tabs = ({
  className,
  content,
  content: { tabs },
  children = [],
  // @ts-ignore
  tabsHeaderContent,
  titleLevel = 2,
  tabFromLocation = false,
  location,
  data,
  dataFetched,
}: TabsProps) => {
  const classWrapper = cx(theme.tabs, className);
  const [active, setActive] = useState<string>(tabs[0].view);
  useEffect(() => {
    const activeTab = tabFromLocation
      ? location &&
        (location.search
          ? new URLSearchParams(location.search).get('tabOpen')
          : tabs[0].view)
      : tabs[0].view;
    setActive(activeTab as string);
  }, [location]);
  useEffect(() => {
    dataFetched
      ? !tabs.filter(tab => tab.view === active && tab.resultsCount).length &&
        setActive(tabs[0].view)
      : setActive(tabs[0].view);
  }, [tabs, dataFetched]);
  let tabItems: JSX.Element[], tabsContents: JSX.Element[];
  tabsContents = children.reduce(
    (accum, child) => {
      if (
        typeof child.props.visible === 'undefined' ||
        child.props.visible === true
      ) {
        const tab = (
          <Tab
            key={child.props.view}
            active={active === child.props.view}
            attributes={
              active === child.props.view
                ? {
                    role: 'tabpanel',
                    'aria-labelledby': `${child.props.view}`,
                    id: `${child.props.view}__tab`,
                  }
                : {
                    role: 'tabpanel',
                    'aria-labelledby': `${child.props.view}`,
                    id: `${child.props.view}__tab`,
                  }
            }
            {...child.props}
          >
            {child.props.children}
          </Tab>
        );
        return [...accum, tab];
      }
      return [...accum];
    },
    [] as JSX.Element[]
  );
  tabItems = tabs.reduce(
    (accum, tab) => {
      if (tabsContents.find(item => item.props.view === tab.view)) {
        const hasResultCount = typeof tab.resultsCount !== 'undefined';

        const tabItem = (
          <Button
            key={tab.view}
            className={cx(theme.tabs__button, 'tabs__button')}
            isToggle
            isDisabled={hasResultCount && !tab.resultsCount}
            role="tab"
            isSelected={active === tab.view}
            onClick={() => setActive(tab.view)}
            viewType={ButtonViewType.icon}
            attributes={
              active === tab.view
                ? {
                    'aria-controls': `${tab.view}__tab`,
                    id: `${tab.view}`,
                  }
                : {
                    'aria-controls': `${tab.view}__tab`,
                    id: `${tab.view}`,
                  }
            }
          >
            {tab.title}
            {hasResultCount ? ` (${tab.resultsCount})` : null}
          </Button>
        );
        return [...accum, tabItem];
      }
      return [...accum];
    },
    [] as JSX.Element[]
  );

  const headerInfo =
    tabsHeaderContent &&
    tabsHeaderContent.contents.find((item: HeaderContent) => {
      return active === item.view;
    });

  let subheading = get(
    get(children.find(child => child.props.view === active), 'props', {}),
    'hasContent',
    {}
  )
    ? get(headerInfo, 'subheading', '')
    : '';

  const processData = (input: string): string => {
    let processed = input;
    data &&
      data.forEach(item => {
        if (processed.includes(item.pattern)) {
          // @ts-ignore
          processed = processed.replace(item.pattern, item.replacement);
        }
      });
    return processed || '';
  };

  return (
    <div className={classWrapper} {...getComponentDataAttrs('tabs', content)}>
      {headerInfo && (
        <div className={cx(theme.tabsHeader, 'tabs-header')}>
          <div
            className={cx(theme.tabsHeader__heading, 'tabs-header__heading')}
          >
            <Text
              // @ts-ignore
              tag={TagName[`h${titleLevel}`]}
              text={processData(headerInfo.heading)}
            />
          </div>
          <div
            className={cx(
              theme.tabsHeader__headingSubheading,
              'tabs-header__heading-subheading'
            )}
          >
            <Text
              // @ts-ignore
              tag={TagName[`h${titleLevel + 1}`]}
              text={processData(subheading)}
            />
          </div>
        </div>
      )}
      <div role="tablist" className={cx(theme.tabs__buttons, 'tabs__buttons')}>
        {tabItems}
      </div>
      <div className={cx(theme.tabs__items, 'tabs__items')}>{tabsContents}</div>
    </div>
  );
};

export default Tabs;
