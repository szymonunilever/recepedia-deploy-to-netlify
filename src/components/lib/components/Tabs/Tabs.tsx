import React, { useState } from 'react';
import { TabsProps, HeaderContent } from './models';
import theme from './Tabs.module.scss';
import cx from 'classnames';
import Button from '../Button';
import { Tab } from './partials';
import { TagName, Text } from 'src/components/lib/components/Text';

export const Tabs = ({
  className,
  content: { tabs },
  children = [],
  // @ts-ignore
  tabsHeaderContent,
}: TabsProps) => {
  const classWrapper = cx(theme.tabs, className);
  const [active, setActive] = useState(tabs[0].view);
  let tabItems: JSX.Element[], tabsContents: JSX.Element[];
  tabItems = tabs.map(tab => (
    <Button
      key={tab.view}
      className="tabs__button"
      isToggle
      role="tab"
      isSelected={active === tab.view}
      toggleExternalManage={true}
      onClick={() => setActive(tab.view)}
      attributes={
        active === tab.view
          ? {
              'aria-controls': `${tab.view}__tab`,
              id: `${tab.view}`,
            }
          : {
              'aria-controls': `${tab.view}__tab`,
              id: `${tab.view}`,
            }
      }
    >
      {tab.title}
    </Button>
  ));
  tabsContents = children.map(child => (
    <Tab
      key={child.props.view}
      active={active === child.props.view}
      attributes={
        active === child.props.view
          ? {
              role: 'tabpanel',
              'aria-labelledby': `${child.props.view}`,
              id: `${child.props.view}__tab`,
            }
          : {
              role: 'tabpanel',
              'aria-labelledby': `${child.props.view}`,
              id: `${child.props.view}__tab`,
            }
      }
      {...child.props}
    >
      {child.props.children}
    </Tab>
  ));

  const headerInfo =
    tabsHeaderContent &&
    tabsHeaderContent.contents.find(
      (item: HeaderContent) => active === item.view
    );

  return (
    <div className={classWrapper} data-componentname="tabs">
      {headerInfo && (
        <div className="tabs-header">
          <div className="tabs-header__heading">
            <Text tag={TagName.h2} text={headerInfo.heading} />
          </div>
          <div className="tabs-header__heading-subheading">
            <Text tag={TagName.h3} text={headerInfo.subheading} />
          </div>
        </div>
      )}
      <div role="tablist" className="tabs__buttons">
        {tabItems}
      </div>
      <div className="tabs__items">{tabsContents}</div>
    </div>
  );
};

export default Tabs;
