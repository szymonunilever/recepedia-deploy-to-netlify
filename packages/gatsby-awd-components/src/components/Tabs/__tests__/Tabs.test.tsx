import { mount } from 'enzyme';
import React from 'react';
import Tabs from '../index';
import toJson from 'enzyme-to-json';

describe('<Tabs />', () => {
  let wrapper: any;
  const tabsHeaderContent = {
    contents: [
      {
        heading: 'Heading',
        subheading: 'subheading',
        view: 'RecipeTabMethods',
        defaultSubheading: 'defaultSubheading',
      },
      {
        heading: 'Heading',
        subheading: '',
        view: 'RecipeTabCooking',
        defaultSubheading: 'defaultSubheading',
      },
    ],
  };

  const props: any = {
    children: [
      {
        view: 'RecipeTabMethods',
        active: true,
        props: {
          view: 'RecipeTabMethods',
          children: <div>RecipeTabMethods</div>,
        },
      },
      {
        view: 'RecipeTabCooking',
        props: {
          view: 'RecipeTabCooking',
          children: <div>RecipeTabCooking</div>,
        },
      },
      {
        view: 'RecipeTabIngredients',
        props: {
          view: 'RecipeTabIngredients',
          children: <div>RecipeTabIngredients</div>,
        },
      },
      {
        view: 'Empty',
        props: {
          view: 'Empty',
          children: <div>Empty</div>,
        },
      },
    ],
    content: {
      tabs: [
        {
          title: 'Recipe',
          view: 'RecipeTabMethods',
        },
        {
          title: 'Cooking',
          view: 'RecipeTabCooking',
        },
        {
          title: 'Ingredients',
          view: 'RecipeTabIngredients',
        },
        {
          title: '',
          view: 'Empty',
        },
      ],
    },
    tabsHeaderContent,
  };

  beforeEach(() => {
    wrapper = mount(<Tabs {...props} />);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('matches the snapshot', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should change tab', () => {
    expect(wrapper.find('#RecipeTabMethods').hasClass('button__selected')).toBe(
      true
    );
    expect(wrapper.find('#RecipeTabCooking').hasClass('button__selected')).toBe(
      false
    );
    wrapper.find('#RecipeTabCooking').simulate('click');
    expect(wrapper.find('#RecipeTabMethods').hasClass('button__selected')).toBe(
      false
    );
    expect(wrapper.find('#RecipeTabCooking').hasClass('button__selected')).toBe(
      true
    );
  });
});
