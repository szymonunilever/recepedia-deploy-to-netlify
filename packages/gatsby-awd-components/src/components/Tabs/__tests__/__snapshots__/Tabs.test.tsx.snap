// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> matches the snapshot 1`] = `
<Component
  content={
    Object {
      "tabs": Array [
        Object {
          "title": "Recipe",
          "view": "RecipeTabMethods",
        },
        Object {
          "title": "Cooking",
          "view": "RecipeTabCooking",
        },
        Object {
          "title": "Ingredients",
          "view": "RecipeTabIngredients",
        },
        Object {
          "title": "",
          "view": "Empty",
        },
      ],
    }
  }
  tabsHeaderContent={
    Object {
      "contents": Array [
        Object {
          "defaultSubheading": "defaultSubheading",
          "heading": "Heading",
          "subheading": "subheading",
          "view": "RecipeTabMethods",
        },
        Object {
          "defaultSubheading": "defaultSubheading",
          "heading": "Heading",
          "subheading": "",
          "view": "RecipeTabCooking",
        },
      ],
    }
  }
>
  <div
    className="tabs"
    data-componentname="tabs"
    data-componentview="tabs"
  >
    <div
      className="tabsHeader tabs-header"
    >
      <div
        className="tabsHeader__heading tabs-header__heading"
      >
        <Text
          tag={1}
          text="Heading"
        >
          <h2>
            Heading
          </h2>
        </Text>
      </div>
      <div
        className="tabsHeader__headingSubheading tabs-header__heading-subheading"
      >
        <Text
          tag={2}
          text="subheading"
        >
          <h3>
            subheading
          </h3>
        </Text>
      </div>
    </div>
    <div
      className="tabs__buttons tabs__buttons"
      role="tablist"
    >
      <Component
        attributes={
          Object {
            "aria-controls": "RecipeTabMethods__tab",
            "id": "RecipeTabMethods",
          }
        }
        className="tabs__button tabs__button"
        isDisabled={false}
        isSelected={true}
        isToggle={true}
        key="RecipeTabMethods"
        onClick={[Function]}
        role="tab"
        viewType={1}
      >
        <button
          aria-controls="RecipeTabMethods__tab"
          aria-selected={true}
          className="buttonIcon button--icon tabs__button tabs__button button__selected"
          data-componentname="button"
          data-componentview="button"
          disabled={false}
          hidden={false}
          id="RecipeTabMethods"
          onClick={[Function]}
          role="tab"
          type="button"
        >
          Recipe
        </button>
      </Component>
      <Component
        attributes={
          Object {
            "aria-controls": "RecipeTabCooking__tab",
            "id": "RecipeTabCooking",
          }
        }
        className="tabs__button tabs__button"
        isDisabled={false}
        isSelected={false}
        isToggle={true}
        key="RecipeTabCooking"
        onClick={[Function]}
        role="tab"
        viewType={1}
      >
        <button
          aria-controls="RecipeTabCooking__tab"
          aria-selected={false}
          className="buttonIcon button--icon tabs__button tabs__button"
          data-componentname="button"
          data-componentview="button"
          disabled={false}
          hidden={false}
          id="RecipeTabCooking"
          onClick={[Function]}
          role="tab"
          type="button"
        >
          Cooking
        </button>
      </Component>
      <Component
        attributes={
          Object {
            "aria-controls": "RecipeTabIngredients__tab",
            "id": "RecipeTabIngredients",
          }
        }
        className="tabs__button tabs__button"
        isDisabled={false}
        isSelected={false}
        isToggle={true}
        key="RecipeTabIngredients"
        onClick={[Function]}
        role="tab"
        viewType={1}
      >
        <button
          aria-controls="RecipeTabIngredients__tab"
          aria-selected={false}
          className="buttonIcon button--icon tabs__button tabs__button"
          data-componentname="button"
          data-componentview="button"
          disabled={false}
          hidden={false}
          id="RecipeTabIngredients"
          onClick={[Function]}
          role="tab"
          type="button"
        >
          Ingredients
        </button>
      </Component>
      <Component
        attributes={
          Object {
            "aria-controls": "Empty__tab",
            "id": "Empty",
          }
        }
        className="tabs__button tabs__button"
        isDisabled={false}
        isSelected={false}
        isToggle={true}
        key="Empty"
        onClick={[Function]}
        role="tab"
        viewType={1}
      >
        <button
          aria-controls="Empty__tab"
          aria-selected={false}
          className="buttonIcon button--icon tabs__button tabs__button"
          data-componentname="button"
          data-componentview="button"
          disabled={false}
          hidden={false}
          id="Empty"
          onClick={[Function]}
          role="tab"
          type="button"
        />
      </Component>
    </div>
    <div
      className="tabs__items tabs__items"
    >
      <Component
        active={true}
        attributes={
          Object {
            "aria-labelledby": "RecipeTabMethods",
            "id": "RecipeTabMethods__tab",
            "role": "tabpanel",
          }
        }
        key="RecipeTabMethods"
        view="RecipeTabMethods"
      >
        <div
          aria-labelledby="RecipeTabMethods"
          className="tabBody"
          hidden={false}
          id="RecipeTabMethods__tab"
          role="tabpanel"
        >
          <div>
            RecipeTabMethods
          </div>
        </div>
      </Component>
      <Component
        active={false}
        attributes={
          Object {
            "aria-labelledby": "RecipeTabCooking",
            "id": "RecipeTabCooking__tab",
            "role": "tabpanel",
          }
        }
        key="RecipeTabCooking"
        view="RecipeTabCooking"
      >
        <div
          aria-labelledby="RecipeTabCooking"
          className="tabBody"
          hidden={true}
          id="RecipeTabCooking__tab"
          role="tabpanel"
        >
          <div>
            RecipeTabCooking
          </div>
        </div>
      </Component>
      <Component
        active={false}
        attributes={
          Object {
            "aria-labelledby": "RecipeTabIngredients",
            "id": "RecipeTabIngredients__tab",
            "role": "tabpanel",
          }
        }
        key="RecipeTabIngredients"
        view="RecipeTabIngredients"
      >
        <div
          aria-labelledby="RecipeTabIngredients"
          className="tabBody"
          hidden={true}
          id="RecipeTabIngredients__tab"
          role="tabpanel"
        >
          <div>
            RecipeTabIngredients
          </div>
        </div>
      </Component>
      <Component
        active={false}
        attributes={
          Object {
            "aria-labelledby": "Empty",
            "id": "Empty__tab",
            "role": "tabpanel",
          }
        }
        key="Empty"
        view="Empty"
      >
        <div
          aria-labelledby="Empty"
          className="tabBody"
          hidden={true}
          id="Empty__tab"
          role="tabpanel"
        >
          <div>
            Empty
          </div>
        </div>
      </Component>
    </div>
  </div>
</Component>
`;
